File: app/src/lib/game/moves.ts
import type { Card, GameMode, PendingEffect, TopCard } from './types.js';
import { isPlayedWhot, isSuitCard } from './guards.js';

export function canPlay(
	card: Card,
	top: TopCard,
	pending: PendingEffect | null,
	mode: GameMode
): boolean {
	// In stack mode with a pending pick, only 2s and 5s can counter
	if (pending?.kind === 'pick' && mode === 'stack') {
		if (!isSuitCard(card)) return false;
		return card.value === 2 || card.value === 5;
	}

	// Whot is always playable (outside of a pending counter window)
	if (card.kind === 'whot') return true;

	// Top is a played whot: must match the called shape
	if (isPlayedWhot(top)) return card.shape === top.calledShape;

	// Standard: match by shape or number
	return card.shape === top.shape || card.value === top.value;
}

export function getValidMoves(
	hand: Card[],
	top: TopCard,
	pending: PendingEffect | null,
	mode: GameMode
): Card[] {
	return hand.filter((card) => canPlay(card, top, pending, mode));
}

File: server/src/game/moves.rs
use crate::game::types::{Card, GameMode, PendingEffect, TopCard};

pub fn can_play(card: Card, top: TopCard, pending: &Option<PendingEffect>, mode: GameMode) -> bool {
    // An opening Whot must be resolved by declaring a shape (a CallShape action),
    // not by playing a card — nothing is playable until then.
    if matches!(pending, Some(PendingEffect::CallShape)) {
        return false;
    }

    // Under a pending penalty, countering is number-locked: only the same
    // number that started it (a 2-chain answered with 2s, a 5-chain with 5s),
    // and only in stack mode. No-stack has no counter — the player must draw.
    if let Some(PendingEffect::Pick { card: counter, .. }) = pending {
        return mode == GameMode::Stack
            && matches!(card, Card::Suit { value, .. } if value == *counter);
    }

    // Whot is always playable outside a counter window
    if matches!(card, Card::Whot) {
        return true;
    }

    let Card::Suit {
        shape: cs,
        value: cv,
    } = card
    else {
        return false;
    };

    match top {
        TopCard::Whot { called_shape } => cs == called_shape,
        TopCard::Suit {
            shape: ts,
            value: tv,
        } => cs == ts || cv == tv,
    }
}

pub fn valid_moves(
    hand: &[Card],
    top: TopCard,
    pending: &Option<PendingEffect>,
    mode: GameMode,
) -> Vec<Card> {
    hand.iter()
        .copied()
        .filter(|&c| can_play(c, top, pending, mode))
        .collect()
}
